### PR TITLE
[trivy] Add option to use Badger DB for caching

### PR DESCRIPTION
**Pull request by davidor**
### What does this PR do?

Adds the option to use [Badger DB](https://github.com/dgraph-io/badger) as a cache for the SBOM collection feature.

Now we're using Bolt which is the default DB supported by Trivy. The main advantage of Badger compared to Bolt for our use case is that Badger allows setting a TTL when storing objects. With Bolt the cache grows until the agent is stopped.

This PR adds a new config option (subject to change until the SBOM collection feature is considered stable):
- `container_image_collection.sbom.cache_ttl` (defaults to 1 hour).

### Additional Notes

The Badger DB is created with the default settings. If performance becomes an issue we can try to tune the settings.

### Describe how to test/QA your changes

Deploy the agent on an environment with containerd and these settings:
- `DD_CONTAINER_IMAGE_COLLECTION_METADATA_ENABLED=true`
- `DD_CONTAINER_IMAGE_COLLECTION_SBOM_ENABLED=true`

Deploy the agent and pull an image. Check the SBOM generation time with `agent workload-list --verbose`. In my case, the datadog/agent:latest images takes a few seconds. Now delete the image, pull it again, and check the generation time again. It should be faster because of the cache.

Also check that `DD_CONTAINER_IMAGE_COLLECTION_SBOM_CACHE_TTL` works as expected. In the process described above, if the time between the first and the second image pull is shorter than the TTL, the SBOM generation time should be fast. Otherwise, it should take the same as the first time.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by vboulineau** [comment]
Considering that this is code is still beta, I would simply use `badger` all the time and put the default configuration as persistent cache.

**Comment by davidor** [comment]
@vboulineau I've made the changes that you suggested.
I also added some call to run the garbage collection because Badger relies on the client to run it (Ref: (https://dgraph.io/docs/badger/get-started/#garbage-collection)

**Comment by davidor** [comment]
Rebased on main to fix conflicts in `go.mod`.

File: pkg/util/trivy/trivy.go [hunk 1]
```diff
@@ -138,7 +138,9 @@ func NewCollector(collectorConfig CollectorConfig) (Collector, error) {
 
 func (c *collector) Close() error {
 	if c.config.ClearCacheOnClose {
-		return c.cache.Clear()
+		if err := c.cache.Clear(); err != nil {
+			return fmt.Errorf("error closing SBOM collector %w", err)
```
[the remark this answers is not in the document]
Comment (davidor) [reply pkg/util/trivy/trivy.go]: Changed :+1:

File: pkg/util/trivy/cache.go [hunk 2]
```diff
@@ -9,16 +9,241 @@
...
+		return nil
+	})
+	if err != nil {
+		return false, nil, fmt.Errorf("error getting missing blobs from Badger cache: %w", err)
+	}
+
+	return missingArtifact, missingBlobIDs, nil
+}
+
+func (cache *BadgerCache) PutArtifact(artifactID string, artifactInfo types.ArtifactInfo) error {
```
Comment (L3n41c) [inline pkg/util/trivy/cache.go:129]: `PutArtifact` and `PutBlob` seems to contain duplicated code.
May I suggest a way to factorize those functions in #15948 ?

File: pkg/util/trivy/cache.go [hunk 3]
```diff
@@ -9,16 +9,241 @@
...
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("error deleting blobs from Badger cache: %w", err)
+	}
+
+	return errs
+}
+
+func (cache *BadgerCache) GetArtifact(artifactID string) (types.ArtifactInfo, error) {
```
Comment (L3n41c) [inline pkg/util/trivy/cache.go:182]: `GetArtifact` and `GetBlob` seems to contain duplicated code.
May I suggest a way to factorize those functions in #15948 ?

**Review by L3n41c — COMMENTED** [review]
I left in #15948 a suggestion to factorize duplicated pieces of code.

**Review by amenasria — APPROVED** [review]
LGTM for Agent Platform owned files !

File: pkg/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go [hunk 4]
```diff
@@ -44,9 +44,7 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
 	trivyConfiguration.ContainerdAccessor = func() (cutil.ContainerdItf, error) {
 		return c.containerdClient, nil
 	}
-	trivyConfiguration.CacheProvider = func() (cache.Cache, error) {
-		return trivy.NewLocalCache(config.Datadog.GetString("container_image_collection.sbom.cache_directory"))
-	}
+	trivyConfiguration.CacheProvider = cacheProvider(true)
```
Comment (davidor) [inline pkg/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:47]: @lebauce If we want to use Badger for the security-agent too we'll need to move this call to the `DefaultCollectorConfig` (called some lines above).

**Review by sgnn7 — APPROVED** [review]
LGTM for `pkg/config/config.go`

**Review by L3n41c — APPROVED** [review]

**Review by lebauce — APPROVED** [review]

**Comment by davidor** [comment]
@lebauce as we agreed, I've added a new commit that configures the security-agent to use Badger as well.

**Review by lebauce — APPROVED** [review]

**Pull request merged**